src/Geometry/Transform.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE KindSignatures #-}
+module Geometry.Transform
+( Transform(..)
+, mkTranslation
+, mkScale
+, mkRotation
+, apply
+, tmap
+, (>>>)
+) where
+
+import Control.Category
+import Control.Lens ((&), (.~))
+import Foreign.Storable
+import GL.Type as GL
+import GL.Uniform
+import Linear.Exts
+import Prelude hiding ((.))
+
+newtype Transform c (a :: * -> *) (b :: * -> *) = Transform { getTransform :: M44 c }
+  deriving (Show, Storable, GL.Type, Uniform)
+
+instance Num c => Category (Transform c) where
+  id = Transform identity
+  Transform a . Transform b = Transform (b !*! a)
+
+mkTranslation :: Num c => V3 c -> Transform c a a
+mkTranslation v = Transform (identity & translation .~ v)
+
+-- FIXME: scaling should introduce a change of units
+mkScale :: Num c => V3 c -> Transform c a b
+mkScale v = Transform (scaled (ext v 1))
```

Comment:
This would help ensure that if we change the solar system units to, say, gigametres, we’ll get errors if any intervening transition is still operating in megametres.